src/tools/rbd/action/Trash.cc

```diff
@@ -0,0 +1,449 @@
+    // image might disappear between rbd.list() and rbd.open(); ignore
+    // that, warn about other possible errors (EPERM, say, for opening
+    // an old-format image, because you need execute permission for the
+    // class method)
+    if (r < 0) {
+      if (r != -ENOENT) {
+        std::cerr << "rbd: error opening " << entry.first << ": "
+                  << cpp_strerror(r) << std::endl;
+      }
+      // in any event, continue to next image
+      continue;
+    }
+
+    // handle second-nth trips through loop
+    parent.clear();
+    r = im.parent_info(&pool, &image, &snap);
+    if (r < 0 && r != -ENOENT)
+      goto out;
+    bool has_parent = false;
+    if (r != -ENOENT) {
+      parent = pool + "/" + image + "@" + snap;
+      has_parent = true;
+    }
+
+    if (im.stat(info, sizeof(info)) < 0) {
+      r = -EINVAL;
+      goto out;
+    }
+
+    uint8_t old_format;
+    im.old_format(&old_format);
+
+    std::list<librbd::locker_t> lockers;
+    bool exclusive;
+    r = im.list_lockers(&lockers, &exclusive, NULL);
+    if (r < 0)
+      goto out;
+    std::string lockstr;
+    if (!lockers.empty()) {
+      lockstr = (exclusive) ? "excl" : "shr";
+    }
+
+    std::string del_source;
+    switch (entry.second.source) {
+      case RBD_TRASH_IMAGE_SOURCE_USER:
+        del_source = "USER";
+        break;
+      case RBD_TRASH_IMAGE_SOURCE_MIRRORING:
+        del_source = "RBD_MIRROR";
+        break;
+    }
+
+    char time_str[128];
+    struct tm *timeinfo;
+    timeinfo = localtime(&entry.second.deletion_time);
+    strftime(time_str, sizeof(time_str), "%F %T%z", timeinfo);
+
+    if (f) {
+      f->open_object_section("image");
+      f->dump_string("id", entry.first);
+      f->dump_string("name", entry.second.name);
+      f->dump_unsigned("size", info.size);
+      if (has_parent) {
+        f->open_object_section("parent");
+        f->dump_string("pool", pool);
+        f->dump_string("image", image);
+        f->dump_string("snapshot", snap);
+        f->close_section();
+      }
+      f->dump_int("format", old_format ? 1 : 2);
+      if (!lockers.empty())
+        f->dump_string("lock_type", exclusive ? "exclusive" : "shared");
+      f->dump_string("del_source", del_source);
+      f->dump_string("deleted_at", time_str);
+      f->dump_string("delete_status",
+                     delete_status(entry.second.deferment_end_time));
+      f->close_section();
+    } else {
+      tbl << entry.first
+          << entry.second.name
+          << stringify(si_t(info.size))
+          << parent
+          << ((old_format) ? '1' : '2')
+          << ""                         // protect doesn't apply to images
+          << lockstr
+          << del_source
+          << time_str
+          << delete_status(entry.second.deferment_end_time)
+          << TextTable::endrow;
+    }
+
+    std::vector<librbd::snap_info_t> snaplist;
+    if (im.snap_list(snaplist) >= 0 && !snaplist.empty()) {
```

Comment:
I don't think we need to list snapshots -- just allow "rbd snap ls --image-id" to function